doc/kuksa_requirementes.md

```diff
@@ -0,0 +1,805 @@
+  - [Provider control and provider capabilities](#provider-control-and-provider-capabilities)
+  - [Control the rate of updates](#control-the-rate-of-updates)
+  - [Differentiate between different providers of the same VSS data](#differentiate-between-different-providers-of-the-same-vss-data)
+  - [Data Aliveness/Availability](#data-alivenessavailability)
+  - [Missing features from `sdv.databroker.v1` in `kuksa.val.v1`](#missing-features-from-sdvdatabrokerv1-in-kuksavalv1)
+- [Exploring a design of a bidirectional streaming API](#exploring-a-design-of-a-bidirectional-streaming-api)
+    - [Design choices](#design-choices)
+        - [Stream](#stream)
+        - [Bidirectional stream](#bidirectional-stream)
+        - [Actuators](#actuators)
+    - [Overview](#overview)
+  - [COVESA topics](#covesa-topics)
+- [Domain requirements](#domain-requirements)
+  - [Interface requirements](#interface-requirements)
+    - [Current Databroker API design](#current-databroker-api-design)
+    - [New Databroker API proposal](#new-databroker-api-proposal)
+- [Appendix](#appendix)
+  - [1. Use Cases](#1-use-cases)
+  - [2. Protobuf API Diagram](#2-protobuf-api-diagram)
+
+# Documentation
+  ### [Terminology](./terminology.md)
+  ### [System Architecture](./system-architecture.md)
+
+# Functional requirements
+
+### As Signal Consumer
+1. As a client I want to have a single service point accessible from my execution environment.
+   
+2. As a client I want to `Get` sensor values or actuator current values of the vehicle and its environment in a uniform manner.
+  - Details:
+    * All sensor and actuator values for which a client has permission to.
+    * Client must provide the paths of the requested signals, which optionally may contain wildcards.
+    * I want to receive exactly all the signals’ values I requested without errors.
+  
+3. As a client I want to `Subscribe` to sensor values or actuator current values when some signal value changed -> what about continuous subscription?
+  - Details:
+    * All sensor and actuator values for which a client has permission to.
+    * Client must provide the path, paths or wildcard paths.
+    * I want to receive exactly all the signals’ values I requested without errors.
+
+4. As a client I want to `Subscribe` and be able to apply a filter to receive a sensor value with an interval of x milliseconds only if the value has changed.
+   
+5.  As a client I want to `GetMetadata` static Metadata from signals.
+  - Details:
+    * All sensors, actuators, and attributes values for which a client has permission to.
+    * Client must provide the path, paths or wildcard.
+
+6.  As a client I want to `Get` attributes values once.
+
+7.  As a client I want to `Actuate` a target value of an actuator. This value has to be forwarded to the provider of this actuator if the provider is available, otherwise the target value is lost. 
```

Comment:
Just to be sure, this is a massive change to the current behavior, right? My expectation in the current databroker would have been that the target value simply stays in the databroker no matter whether someone consumed the data or not. 

Now, it seems that you only want to forward target values if there is already someone subscribed to this value who can perform the actuation (aka actuation provider). This requires a 1:1 connection between the target of a signal and an already subscribed provider. 

I am a bit afraid that we limit ourselves too much here. For instance, how about cases where a third-party is interested in what actuation if currently going on (or failed)? I could also think about cases, where the target value is still valid even there was no provider in the moment it was set. Here the provider could check the timestamp and decide for itself whether the actuation shall be performed or not.

Reply:
I was discussing with Rafael and John about specifying a lifetime of a value (also for sensor and current actuator values). This could be also used for target values: If the provider is not "picking" a target value up within its lifetime, it is dropped. That means, that a short "outage" of a provider does not lead automatically to a lost actuation request, but target values set "some hours" before also don't lead to no longer expect actuation. 

John's counter-arg was that the info about temporary missing provider would be given back to the client (via signal specific failure) and then it would be the clients task to retry once the provider "is back" (and the actuation still is wanted).

Reply:
Forgot to state, that I think John's point is valid. The "pretended convenience" of storing the target value in the broker for a certain time is a false one: The consumer must be able to handle outages of the broker itself, anyway. Therefore, it's best to implement the retry on consumer side. This could be a convenience functionality offered by the consumer SDK/library.

Reply:
The "pretended convenience" is a good point. 
It still means that the consumer who wants to trigger the actuation needs to find the moment in time when the actuation provider is connected. 
With defining a lifetime of how long the actuation target is valid when setting the target value, one could ease this is a bit. But I am not sure whether we will see many useful actuation provider which really come and leave that frequently so that this becomes relevant.


Would this mean that the databroker only brokers sensor values and for actuation more acts as a router and translator?

Reply:
> Would this mean that the databroker only brokers sensor values and for actuation more acts as a router and translator?

Yes. Or put another way:
- Signal values (data) are cached.
- Actuation requests (requests, possibly expecting a response) are not cached.

Even if we decide that actuation requests should not return a response, it is very hard / impossible to formulate a consistent cache invalidation policy for actuation requests if we were to model them as "setting a field in a database", i.e. "target values".

It does not play well with actuators that can be controlled not only through databroker - from other ECUS over CAN or directly connected physical buttons - which I expect would be a pretty standard situation. When that happens, the target value is outdated and pretty much useless.

At the time of writing, I don't know a single provider implementation that checks the timestamp of a target value or at any point "clears" them. Neither are any consumers that I'm aware of. They essentially live forever. I think it's a lot easier to write robust providers (and consumers) by modelling actuation as a request / response thing.